feat(nuxt): add option to disable modules from layers

janniksohn opened the pull request:
## 🔗 Linked issue

resolves https://github.com/nuxt/nuxt/issues/32949

## 📚 Description

## Overview

This feature introduces a new `disabledModules` configuration option that allows Nuxt projects to selectively disable modules inherited from extended layers.

## Problem

When using Nuxt layers, all modules defined in those layers are automatically installed in the consuming project. This can be problematic in several scenarios:

1. **Unwanted dependencies**: A layer might include modules that the consuming project doesn't need, adding unnecessary bundle size and build time.

2. **Module conflicts**: A layer module might conflict with a different module or custom implementation in the consuming project.

3. **Environment-specific requirements**: Certain modules from a layer may not be appropriate for all deployment environments (e.g., analytics modules in development).

4. **Customization limitations**: Previously, there was no way to "opt-out" of specific layer modules without forking the entire layer or restructuring it.

5. **Gradual migration**: When migrating away from certain modules, teams need a way to disable them in specific projects while keeping them in shared layers.

### Example

Consider a company with a shared base layer:

```ts
// shared-layer/nuxt.config.ts
export default defineNuxtConfig({
  modules: [
    '@nuxt/image',
    '@pinia/nuxt',
    '@sentry/nuxt/module',
  ]
})
```

A new project extends this layer but:
- Uses a different image optimization solution
- Doesn't need analytics during early development
- Has conflicting Sentry configuration

Previously, there was no clean way to exclude these modules without modifying the shared layer itself.

## Solution

The `disabledModules` option provides a declarative way to exclude specific modules from layers:

```ts
// project/nuxt.config.ts
export default defineNuxtConfig({
  extends: ['../shared-layer'],
  disabledModules: ['@nuxt/image', '@sentry/nuxt/module']
})
```

## How It Works

### Module Resolution Changes

1. **Layer module tracking**: During module resolution, Nuxt now tracks all module names defined in extended layers (stored in `_layerModules`).

2. **Disabled module filtering**: When processing layer modules, the resolver checks against the `disabledModules` list and skips matching modules.

3. **Root project protection**: Modules defined directly in the root project's `nuxt.config.ts` are never disabled, even if listed in `disabledModules`. This prevents accidental self-disabling.

4. **Validation warnings**: If a module name in `disabledModules` doesn't match any layer module, a warning is logged to help catch typos or outdated configurations.

### Type Generation

A new type template (`disabled-modules.d.ts`) is generated that provides:
- A union type of all available layer module names
- Type-safe autocomplete for the `disabledModules` option in `nuxt.config.ts`

## Key Design Decisions

1. **String-based matching**: Modules are matched by their string identifier (package name or path), making configuration straightforward and readable.

2. **Layer-only scope**: Only modules from layers can be disabled. Root project modules are always installed, ensuring developers maintain full control over their direct dependencies.

3. **Warning on mismatch**: Rather than silently ignoring invalid entries, the system warns about disabled modules not found in any layer, helping maintain clean configurations.

4. **No runtime impact**: This is purely a build-time feature. Disabled modules are simply not installed, resulting in no runtime overhead.

## Files Changed

| File | Purpose |
|------|---------|
| `packages/schema/src/types/schema.ts` | Added `disabledModules` and `_layerModules` type definitions |
| `packages/schema/src/config/common.ts` | Added resolver for `disabledModules` option |
| `packages/schema/src/config/internal.ts` | Added internal `_layerModules` option |
| `packages/nuxt/src/core/nuxt.ts` | Implemented module filtering logic in `resolveModules` |
| `packages/nuxt/src/core/templates.ts` | Added type template for autocomplete support |

## Testing

tests cover:
- Loading all modules when `disabledModules` is not set
- Disabling single and multiple modules from layers
- Ensuring root project modules cannot be disabled
- Warning when disabled modules are not found


<!----------------------------------------------------------------------
Before creating the pull request, please make sure you do the following:

- Check that there isn't already a PR that solves the problem the same way. If you find a duplicate, please help us reviewing it.
- Read the contribution docs at https://nuxt.com/docs/4.x/community/contribution
- Ensure that PR title follows conventional commits (https://www.conventionalcommits.org)
- Update the corresponding documentation if needed.
- Include relevant tests that fail without this PR but pass with it.

Thank you for contributing to Nuxt!
----------------------------------------------------------------------->

janniksohn committed:
feat(nuxt): add  option to disable modules from layers

This allows selectively disabling modules that are defined in extended layers\
without affecting the root project's modules.

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

danielroe reviewed (commented):
I would like to implement this by supporting passing `false` to the module options.

we need to special-case 'pages' and maybe(?) 'devtools'

eg. `fonts: false` would disable the fonts module

janniksohn committed:
refactor(kit): simplify disabled module handling with configKey approach

janniksohn commented:
```TS
  // Before
  export default defineNuxtConfig({
    disabledModules: ['@nuxtjs/i18n']
  })

  // After
  export default defineNuxtConfig({
    i18n: false
  })
```

  Key changes:
  - Modules from layers are marked with `_fromLayer` flag during resolution
  - Disabled check happens in `installModules` after module metadata is loaded (to access configKey)
  - Disabled modules are still registered (for type generation) but their `setup()` is not executed
  - Extracted `registerInstalledModule` helper to deduplicate registration logic

janniksohn committed:
fix(nuxt): add false type to module config for disable support

Allow `false` as a valid type for module configuration options, enabling type-safe module disabling from layers.

danielroe commented on packages/nuxt/src/core/nuxt.ts:
```diff
@@ -923,12 +923,20 @@ async function resolveModules (nuxt: Nuxt) {
   const configs = nuxt.options._layers.map(layer => layer.config).reverse()
   for (const config of configs) {
+    // Check if this is the root project (modules from root project cannot be disabled)
+    const isRootProject = config.rootDir === nuxt.options.rootDir
```
I don't think we need to track modules from layers - any module should be able to be disabled

danielroe commented on packages/schema/src/config/common.ts:
```diff
@@ -157,6 +157,19 @@ export default defineResolvers({
+      }
+      return disabledModules
+    },
+  },
```
```suggestion
```

janniksohn committed:
refactor(kit): simplify module disable mechanism

janniksohn commented:
### Chagnes
- **Removed leftover code:** The `disabledModules` array config aswell as `_layerModules` and `_fromLayer` flag, that are no longer needed, were accidentally left in => now removed
- **Extended disable scope:** Any module can now be disabled via `<configKey>: false`, not just modules from layers
- **Core module protection:** Added `ignoredConfigKeys` set for `pages`, `components`, and `devtools` which handle `false` as a valid config value internally

danielroe committed:
Merge remote-tracking branch 'origin/main' into layer-module-disable

danielroe committed:
fix(nuxt): ensure false is always valid even if module types can't be inferred

danielroe committed:
chore: simplify

danielroe committed:
refactor: slightly clean up

danielroe committed:
test: use config keys in test

danielroe committed:
fix: move false to correct location

danielroe committed:
Merge remote-tracking branch 'origin/main' into layer-module-disable

danielroe committed:
chore: add telemetry to allowlist

danielroe committed:
chore: test explicitly for `false`

danielroe reviewed (approved):
thank you for this! ❤️ 

I took the opportunity to clean up some of the kit code (preexisting the PR). hope that was ok! 🙏
